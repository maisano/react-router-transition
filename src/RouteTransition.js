import React, { cloneElement, createElement, Component } from 'react';
import TransitionMotion from 'react-motion/lib/TransitionMotion';
import PropTypes from 'prop-types';

import ensureSpring from './ensureSpring';

const identity = val => val;

class RouteTransition extends Component {
  static defaultProps = {
    component: 'span',
    runOnMount: false,
    mapStyles: identity,
  };

  static propTypes = {
    className: PropTypes.string,
    component: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.element,
      PropTypes.string,
    ]),
    atEnter: PropTypes.object.isRequired,
    atActive: PropTypes.object.isRequired,
    atLeave: PropTypes.object.isRequired,
    mapStyles: PropTypes.func.isRequired,
    runOnMount: PropTypes.bool.isRequired,
  };

  getDefaultStyles() {
    if (!this.props.runOnMount) {
      return null;
    }

    if (!this.props.children) {
      return [];
    }

    return [
      {
        key: this.props.children.key,
        data: this.props.children,
        style: this.props.atEnter,
      },
    ];
  }

  // there's only ever one route mounted at a time,
  // so just return the current match
  getStyles() {
    if (!this.props.children) {
      return [];
    }

    return [
      {
        key: this.props.children.key,
        data: this.props.children,
        style: ensureSpring(this.props.atActive),
      },
    ];
  }

  willEnter = () => this.props.atEnter;

  willLeave = () => ensureSpring(this.props.atLeave);

  renderRoute = config => {
    const props = {
      style: this.props.mapStyles(config.style),
      key: config.key,
    };

    return this.props.component !== false
      ? createElement(this.props.component, props, config.data)
      : cloneElement(config.data, props);
  };

  renderRoutes = interpolatedStyles =>
    (<div className={this.props.className}>
      {interpolatedStyles.map(this.renderRoute)}
    </div>);

  render() {
    return (
      <TransitionMotion
        defaultStyles={this.getDefaultStyles()}
        styles={this.getStyles()}
        willEnter={this.willEnter}
        willLeave={this.willLeave}
      >
        {this.renderRoutes}
      </TransitionMotion>
    );
  }
}

export default RouteTransition;
