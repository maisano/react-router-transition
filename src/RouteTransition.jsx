import React, { cloneElement, createElement, Component } from 'react';
import TransitionMotion from 'react-motion/lib/TransitionMotion';
import PropTypes from 'prop-types';

import ensureSpring from './ensureSpring';

class RouteTransition extends Component {
  constructor(props) {
    super(props);
    this.renderRoute = this.renderRoute.bind(this);
    this.renderRoutes = this.renderRoutes.bind(this);
    this.willEnter = this.willEnter.bind(this);
    this.willLeave = this.willLeave.bind(this);
  }

  getDefaultStyles() {
    if (!this.props.runOnMount) {
      return null;
    }

    if (!this.props.children) {
      return [];
    }

    return [
      {
        key: this.props.pathname,
        data: this.props.children,
        style: this.props.atEnter,
      },
    ];
  }

  // there's only ever one route mounted at a time,
  // so just return the current match
  getStyles() {
    if (!this.props.children) {
      return [];
    }

    // TODO: maybe access route path from children for pathname?
    return [
      {
        key: this.props.pathname,
        data: this.props.children,
        style: ensureSpring(this.props.atActive),
      },
    ];
  }

  willEnter() {
    return this.props.atEnter;
  }

  willLeave() {
    return ensureSpring(this.props.atLeave);
  }

  renderRoute(config) {
    const props = {
      style: this.props.mapStyles(config.style),
      key: config.key,
    };

    return this.props.component
      ? createElement(this.props.component, props, config.data)
      : cloneElement(config.data, props);
  }

  renderRoutes(interpolatedStyles) {
    return (
      <div className={this.props.className} style={this.props.style}>
        {interpolatedStyles.map(this.renderRoute)}
      </div>
    );
  }

  render() {
    return (
      <TransitionMotion
        defaultStyles={this.getDefaultStyles()}
        styles={this.getStyles()}
        willEnter={this.willEnter}
        willLeave={this.willLeave}
      >
        {this.renderRoutes}
      </TransitionMotion>
    );
  }
}

RouteTransition.defaultProps = {
  component: 'div',
  runOnMount: true,
  mapStyles: val => val,
};

RouteTransition.propTypes = {
  className: PropTypes.string,
  component: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  pathname: PropTypes.string.isRequired,
  atEnter: PropTypes.object.isRequired,
  atActive: PropTypes.object.isRequired,
  atLeave: PropTypes.object.isRequired,
  mapStyles: PropTypes.func,
  runOnMount: PropTypes.bool,
  style: PropTypes.object,
};

export default RouteTransition;
