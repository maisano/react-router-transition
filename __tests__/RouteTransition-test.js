import React from 'react';
import { shallow } from 'enzyme';
import RouteTransition from '../src/RouteTransition';

test('RouteTransition mounts', () => {
  // Render a checkbox with label in the document
  const routeTransition = shallow(
    <RouteTransition
      atActive={{}}
      atLeave={{}}
      atEnter={{}}
    />
  );

  expect(routeTransition.html()).toMatchSnapshot();
});
