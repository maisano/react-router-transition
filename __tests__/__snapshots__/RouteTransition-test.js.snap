// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteTransition mounts 1`] = `"<div></div>"`;
